### PR TITLE
feature: support Dameng database

### LINKED ISSUE #4014: seata框架 是否支持达梦（DM8）数据库
官网说明 AT模式不支持 达梦，那么TCC模式是否可以实现呢

### PR DESCRIPTION
**Pull request by iquanzhan**
### Ⅰ. Describe what this PR did
support DM DataBase

### Ⅱ. Does this pull request fix one issue?

fixes #4014 

### Ⅳ. Describe how to verify it
download DM,and add table
sql:
```sql
CREATE TABLE "UNDO_LOG"
(
    "ID" BIGINT IDENTITY(1, 1) NOT NULL,
    "BRANCH_ID" BIGINT NOT NULL,
    "XID" VARCHAR(100) NOT NULL,
    "CONTEXT" VARCHAR(128) NOT NULL,
    "ROLLBACK_INFO" BLOB NOT NULL,
    "LOG_STATUS" INT NOT NULL,
    "LOG_CREATED" TIMESTAMP(0) NOT NULL,
    "LOG_MODIFIED" TIMESTAMP(0) NOT NULL,
    "EXT" VARCHAR(100),
    NOT CLUSTER PRIMARY KEY("ID"),
    CONSTRAINT "UX_UNDO_LOG" UNIQUE("XID", "BRANCH_ID")
) STORAGE(ON "MAIN", CLUSTERBTR) ;

CREATE UNIQUE  INDEX "PRIMARY" ON "SEATA_ORDER"."UNDO_LOG"("ID" ASC) STORAGE(ON "MAIN", CLUSTERBTR) ;
```

### Ⅴ. Special notes for reviews
add support DM DataBase

### CODE REVIEW

**Review by l81893521 — COMMENTED** [review]
CI failed. Please check the error detail in https://github.com/seata/seata/pull/3672/checks

File: sqlparser/seata-sqlparser-druid/src/main/java/io/seata/sqlparser/druid/dm/DmDeleteRecognizer.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import io.seata.common.exception.NotSupportYetException;
+import io.seata.sqlparser.ParametersHolder;
+import io.seata.sqlparser.SQLDeleteRecognizer;
+import io.seata.sqlparser.SQLType;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * The type oracle delete recognizer.
```
Comment (l81893521) [inline sqlparser/seata-sqlparser-druid/src/main/java/io/seata/sqlparser/druid/dm/DmDeleteRecognizer.java:34]: The type db delete recognizer.

**Review by l81893521 — COMMENTED** [review]
Please create the sql script in /script/client/at/db/dm.sql

**Comment by a364176773** [comment]
> 代码已重新提交并优化。
> 使用达梦数据库的分布式事务示例：https://gitee.com/iquanzhan/seata-parent

方便给我私发一下你的钉钉号吗？我邮箱364176773@qq.com

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/dm/DmUndoDeleteExecutor.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import io.seata.rm.datasource.undo.AbstractUndoExecutor;
+import io.seata.rm.datasource.undo.SQLUndoLog;
+import io.seata.sqlparser.util.JdbcConstants;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * The type oracle undo delete executor.
```
Comment (l81893521) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/dm/DmUndoDeleteExecutor.java:33]: dm

**Comment by Musknine** [comment]
image
something like this PR #4005.  I thought there are some bugs.     only the unique index and all the unique index's columes same as primary key columes

**Review by Musknine — COMMENTED** [review]
image
there is a bug. only the unique index and all the unique index's columes same as primary key columes,it belongs to primary key     it's something like this PR #4005

**Comment by Musknine** [comment]
@iquanzhan    请问下，支持DM这个pr， 你们有用在生产跑了吗？

**Comment by kzq1205** [comment]
使用达梦数据库的分布式事务示例：https://gitee.com/iquanzhan/seata-parent。
您好，这个示例目前只有仓库成员有访问权限，请问能开放下这个仓库的公开访问权限吗？

**Comment by pengten** [comment]
> 使用达梦数据库的分布式事务示例：[https://gitee.com/iquanzhan/seata-parent。](https://gitee.com/iquanzhan/seata-parent%E3%80%82) 您好，这个示例目前只有仓库成员有访问权限，请问能开放下这个仓库的公开访问权限吗？

这是私人仓库吧，去联系仓库owner

**Comment by MoyaiTing** [comment]
预计什么时候合并到2.0.0？

**Comment by iquanzhan** [comment]
因为内网开发原因，不太方便经常登录github。关于Seata对达梦数据库的支持，建议大家通过SPI机制外挂实现对达梦数据库的支持，目前已通过此方式用于生产。基础版本为：[seata-dm-spring-boot-starter](https://gitee.com/iquanzhan/seata-dm-support/tree/master/seata-dm-spring-boot-starter)

File: script/client/at/db/dm.sql [hunk 3]
```diff
@@ -0,0 +1,15 @@
+CREATE TABLE "UNDO_LOG"
+(
+"ID" BIGINT IDENTITY(1, 1) NOT NULL,
+"BRANCH_ID" BIGINT NOT NULL,
+"XID" VARCHAR(100) NOT NULL,
```
[the remark this answers is not in the document]
Comment (a364176773) [reply script/client/at/db/dm.sql]: done

**Review by wangliang181230 — APPROVED** [review]
LGTM

**Comment by teng-jy** [comment]
10:27:11.956  INFO --- [     RetryRollbacking_1_1] [server.coordinator.DefaultCore] [    doGlobalRollback]  [192.168.2.50:8091:7800661551294558767] : Rollback global transaction successfully, xid = 192.168.2.50:8091:7800661551294558767.
10:29:09.177  INFO --- [     batchLoggerPrint_1_1] [ocessor.server.BatchLogHandler] [                 run]  [] : receive msg[single]: GlobalBeginRequest{transactionName='save(com.beichuanghui.bch.base.vo.save.user.BaseEmployeeSaveVO)', timeout=60000}, clientIp: 192.168.2.168, vgroup: default_tx_group
10:29:09.179  INFO --- [verHandlerThread_1_49_500] [coordinator.DefaultCoordinator] [       doGlobalBegin]  [192.168.2.50:8091:7800661551294558907] : Begin new global transaction applicationId: beichuanghui-base-server,transactionServiceGroup: default_tx_group, transactionName: save(com.beichuanghui.bch.base.vo.save.user.BaseEmployeeSaveVO),timeout:60000,xid:192.168.2.50:8091:7800661551294558907
10:29:09.183  INFO --- [     batchLoggerPrint_1_1] [ocessor.server.BatchLogHandler] [                 run]  [] : result msg[single]: GlobalBeginResponse{xid='192.168.2.50:8091:7800661551294558907', extraData='null', resultCode=Success, msg='null'}, clientIp: 192.168.2.168, vgroup: default_tx_group
10:29:09.289  INFO --- [     batchLoggerPrint_1_1] [ocessor.server.BatchLogHandler] [                 run]  [] : receive msg[merged]: BranchRegisterRequest{xid='192.168.2.50:8091:7800661551294558907', branchType=AT, resourceId='jdbc:dm://192.168.2.30:5236/beichuanghui_ds_c_defaults', lockKey='def_user:397953020899360832;def_user_tenant_rel:397953020899360833', applicationData='{"autoCommit":false,"skipCheckLock":true}'}, clientIp: 192.168.2.168, vgroup: default_tx_group
10:29:09.293  INFO --- [nPool.commonPool-worker-3] [erver.coordinator.AbstractCore] [bda$branchRegister$0]  [192.168.2.50:8091:7800661551294558907] : Register branch successfully, xid = 192.168.2.50:8091:7800661551294558907, branchId = 7800661551294558909, resourceId = jdbc:dm://192.168.2.30:5236/beichuanghui_ds_c_defaults ,lockKeys = def_user:397953020899360832;def_user_tenant_rel:397953020899360833
10:29:09.296  INFO --- [     batchLoggerPrint_1_1] [ocessor.server.BatchLogHandler] [                 run]  [] : result msg[merged]: BranchRegisterResponse{branchId=7800661551294558909, resultCode=Success, msg='null'}, clientIp: 192.168.2.168, vgroup: default_tx_group
10:29:09.370  INFO --- [     batchLoggerPrint_1_1] [ocessor.server.BatchLogHandler] [                 run]  [] : receive msg[merged]: BranchRegisterRequest{xid='192.168.2.50:8091:7800661551294558907', branchType=AT, resourceId='jdbc:dm://192.168.2.30:5236/beichuanghui_ds_c_defaults', lockKey='def_user:397953020899360832;undo_log:30;def_user_tenant_rel:397953020899360833', applicationData='{"autoCommit":false,"skipCheckLock":true}'}, clientIp: 192.168.2.168, vgroup: default_tx_group
10:29:09.373 ERROR --- [nPool.commonPool-worker-3] [ption.AbstractExceptionHandler] [eptionHandleTemplate]  [192.168.2.50:8091:7800661551294558907] : Catch TransactionException while do RPC, request: BranchRegisterRequest{xid='192.168.2.50:8091:7800661551294558907', branchType=AT, resourceId='jdbc:dm://192.168.2.30:5236/beichuanghui_ds_c_defaults', lockKey='def_user:397953020899360832;undo_log:30;def_user_tenant_rel:397953020899360833', applicationData='{"autoCommit":false,"skipCheckLock":true}'}
==>
io.seata.core.exception.TransactionException: branch register request failed. xid=192.168.2.50:8091:7800661551294558907, msg=违反表[LOCK_TABLE]唯一性约束
	at io.seata.server.AbstractTCInboundHandler$4.execute(AbstractTCInboundHandler.java:187) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.server.AbstractTCInboundHandler$4.execute(AbstractTCInboundHandler.java:179) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.core.exception.AbstractExceptionHandler.exceptionHandleTemplate(AbstractExceptionHandler.java:131) ~[seata-core-1.7.1-SNAPSHOT.jar:1.7.1-SNAPSHOT]
	at io.seata.server.AbstractTCInboundHandler.handle(AbstractTCInboundHandler.java:179) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.core.protocol.transaction.BranchRegisterRequest.handle(BranchRegisterRequest.java:136) ~[seata-core-1.7.1-SNAPSHOT.jar:1.7.1-SNAPSHOT]
	at io.seata.server.coordinator.DefaultCoordinator.onRequest(DefaultCoordinator.java:523) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.core.rpc.processor.server.ServerOnRequestProcessor.handleRequestsByMergedWarpMessage(ServerOnRequestProcessor.java:288) ~[seata-core-1.7.1-SNAPSHOT.jar:1.7.1-SNAPSHOT]
	at io.seata.core.rpc.processor.server.ServerOnRequestProcessor.lambda$onRequestMessage$1(ServerOnRequestProcessor.java:178) ~[seata-core-1.7.1-SNAPSHOT.jar:1.7.1-SNAPSHOT]
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1692) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183) ~[na:na]
Caused by: io.seata.common.exception.StoreException: 违反表[LOCK_TABLE]唯一性约束
	at io.seata.server.storage.db.lock.LockStoreDataBaseDAO.acquireLock(LockStoreDataBaseDAO.java:199) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.server.storage.db.lock.DataBaseLocker.acquireLock(DataBaseLocker.java:64) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.server.lock.AbstractLockManager.acquireLock(AbstractLockManager.java:65) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.server.session.BranchSession.lock(BranchSession.java:287) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.server.transaction.at.ATCore.branchSessionLock(ATCore.java:77) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.server.coordinator.AbstractCore.lambda$branchRegister$0(AbstractCore.java:83) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.server.storage.db.session.DataBaseSessionManager.lockAndExecute(DataBaseSessionManager.java:191) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.server.session.SessionHolder.lockAndExecute(SessionHolder.java:367) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.server.coordinator.AbstractCore.branchRegister(AbstractCore.java:77) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.server.coordinator.DefaultCore.branchRegister(DefaultCore.java:103) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.server.coordinator.DefaultCoordinator.doBranchRegister(DefaultCoordinator.java:294) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.server.AbstractTCInboundHandler$4.execute(AbstractTCInboundHandler.java:184) ~[classes!/:1.7.1-SNAPSHOT]
	... 14 common frames omitted
Caused by: java.sql.BatchUpdateException: 违反表[LOCK_TABLE]唯一性约束
	at dm.jdbc.driver.DBError.throwBatchUpdateException(DBError.java:728) ~[DmJdbcDriver18-8.1.2.192.jar:- 8.1.2.192 - Production]
	at dm.jdbc.a.a.a(DBAccess.java:910) ~[DmJdbcDriver18-8.1.2.192.jar:- 8.1.2.192 - Production]
	at dm.jdbc.driver.DmdbPreparedStatement.do_executeLargeBatch(DmdbPreparedStatement.java:654) ~[DmJdbcDriver18-8.1.2.192.jar:- 8.1.2.192 - Production]
	at dm.jdbc.driver.DmdbPreparedStatement.do_executeBatch(DmdbPreparedStatement.java:601) ~[DmJdbcDriver18-8.1.2.192.jar:- 8.1.2.192 - Production]
	at dm.jdbc.driver.DmdbPreparedStatement.executeBatch(DmdbPreparedStatement.java:1623) ~[DmJdbcDriver18-8.1.2.192.jar:- 8.1.2.192 - Production]
	at com.alibaba.druid.pool.DruidPooledPreparedStatement.executeBatch(DruidPooledPreparedStatement.java:565) ~[druid-1.2.6.jar:1.2.6]
	at io.seata.server.storage.db.lock.LockStoreDataBaseDAO.doAcquireLocks(LockStoreDataBaseDAO.java:371) ~[classes!/:1.7.1-SNAPSHOT]
	at io.seata.server.storage.db.lock.LockStoreDataBaseDAO.acquireLock(LockStoreDataBaseDAO.java:187) ~[classes!/:1.7.1-SNAPSHOT]
	... 25 common frames omitted
<==

10:29:09.376  INFO --- [     batchLoggerPrint_1_1] [ocessor.server.BatchLogHandler] [                 run]  [] : result msg[merged]: BranchRegisterResponse{branchId=0, resultCode=Failed, msg='TransactionException[branch register request failed. xid=192.168.2.50:8091:7800661551294558907, msg=违反表[LOCK_TABLE]唯一性约束]'}, clientIp: 192.168.2.168, vgroup: default_tx_group
10:29:09.382  INFO --- [     batchLoggerPrint_1_1] [ocessor.server.BatchLogHandler] [                 run]  [] : receive msg[merged]: BranchReportRequest{xid='192.168.2.50:8091:7800661551294558907', branchId=7800661551294558909, resourceId='null', status=PhaseOne_Failed, applicationData='null', branchType=AT}, clientIp: 192.168.2.168, vgroup: default_tx_group
10:29:09.385  INFO --- [nPool.commonPool-worker-3] [erver.coordinator.AbstractCore] [        branchReport]  [192.168.2.50:8091:7800661551294558907] : Report branch status successfully, xid = 192.168.2.50:8091:7800661551294558907, branchId = 7800661551294558909
10:29:09.387  INFO --- [     batchLoggerPrint_1_1] [ocessor.server.BatchLogHandler] [                 run]  [] : result msg[merged]: BranchReportResponse{resultCode=Success, msg='null'}, clientIp: 192.168.2.168, vgroup: default_tx_group
10:29:09.391  INFO --- [nPool.commonPool-worker-3] [erver.coordinator.AbstractCore] [        branchReport]  [192.168.2.50:8091:7800661551294558907] : Report branch status successfully, xid = 192.168.2.50:8091:7800661551294558907, branchId = 7800661551294558909
10:29:09.393  INFO --- [     batchLoggerPrint_1_1] [ocessor.server.BatchLogHandler] [                 run]  [] : receive msg[merged]: BranchReportRequest{xid='192.168.2.50:8091:7800661551294558907', branchId=7800661551294558909, resourceId='null', status=PhaseOne_Failed, applicationData='null', branchType=AT}, clientIp: 192.168.2.168, vgroup: default_tx_group
10:29:09.393  INFO --- [     batchLoggerPrint_1_1] [ocessor.server.BatchLogHandler] [                 run]  [] : result msg[merged]: BranchReportResponse{resultCode=Success, msg='null'}, clientIp: 192.168.2.168, vgroup: default_tx_group
10:29:09.398  INFO --- [     batchLoggerPrint_1_1] [ocessor.server.BatchLogHandler] [                 run]  [] : receive msg[single]: GlobalRollbackRequest{xid='192.168.2.50:8091:7800661551294558907', extraData='null'}, clientIp: 192.168.2.168, vgroup: default_tx_group
10:29:09.402  INFO --- [rverHandlerThread_1_4_500] [server.coordinator.DefaultCore] [    doGlobalRollback]  [192.168.2.50:8091:7800661551294558907] : Rollback global transaction successfully, xid = 192.168.2.50:8091:7800661551294558907.
10:29:09.403  INFO --- [     batchLoggerPrint_1_1] [ocessor.server.BatchLogHandler] [                 run]  [] : result msg[single]: GlobalRollbackResponse{globalStatus=Rollbacked, resultCode=Success, msg='null'}, clientIp: 192.168.2.168, vgroup: default_tx_group

出现了违反表结构的错误

**Comment by a364176773** [comment]
> 

这个是dameng的driver有bug吧,这个时候应该抛出的是SQLIntegrityConstraintViolationException,而不是BatchUpdateException

**Comment by a364176773** [comment]
@teng-jy 你能联系到dameng的技术人员吗,反馈下这个问题

**Review by a364176773 — APPROVED** [review]
LGTM

**Review by slievrly — APPROVED** [review]
LGTM

**Comment by jinyulei0710** [comment]
any plan merge to 1.6.1?

**Pull request merged**